x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,313 @@
+
+    private static final ParseField QUERY_FIELD = new ParseField("query");
+
+    private final String fieldName;
+    private final String query;
+
+    private SetOnce<InferenceServiceResults> inferenceResultsSupplier;
+    private InferenceServiceResults inferenceResults;
+
+    public SemanticQueryBuilder(String fieldName, String query) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a fieldName");
+        }
+        if (query == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a " + QUERY_FIELD.getPreferredName() + " value");
+        }
+        this.fieldName = fieldName;
+        this.query = query;
+    }
+
+    public SemanticQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.query = in.readString();
+        this.inferenceResults = in.readOptionalNamedWriteable(InferenceServiceResults.class);
+        if (this.inferenceResults != null) {
+            // The supplier is generally not used after the results are set, but set the supplier to maintain equality when serializing
+            // & deserializing
+            this.inferenceResultsSupplier = new SetOnce<>(this.inferenceResults);
+        }
+    }
+
+    private SemanticQueryBuilder(
+        SemanticQueryBuilder other,
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier,
+        InferenceServiceResults inferenceResults
+    ) {
+        this.fieldName = other.fieldName;
+        this.query = other.query;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.inferenceResultsSupplier = inferenceResultsSupplier;
+        this.inferenceResults = inferenceResults;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.SEMANTIC_TEXT_FIELD_ADDED;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (inferenceResultsSupplier != null && inferenceResults == null) {
+            throw new IllegalStateException("Inference results supplier is set, but inference results is null. Missing a rewriteAndFetch?");
+        }
+        out.writeString(fieldName);
+        out.writeString(query);
+        out.writeOptionalNamedWriteable(inferenceResults);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.field(QUERY_FIELD.getPreferredName(), query);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
```

Comment:
I want to refactor this method to make it a bit more readable. It's kind of sprawly right now and I think it would help to factor out each rewrite step into its own private method.